### DISCUSSION TITLE
Implementation of subset-aware function layers

### DISCUSSION THREAD

**Discussion by rafdouglas** [UNDP-Data/geohub#1568] [Ideas]
**UNDP - SDGi – GeoHub**

**Internal note**

# Proposal: implementation of
 subset-aware function layers

_March 16 __th__ , 2023_

## ​Background

Currently we imported ~250 vector files into GeoHub.

Internally, PostgreSQL-side, these files are accessed as **views** which connect the flat data with geographical `admin0` features.

This step was important to assess the data we have and get a first visualization.

However, some shortcomings emerged, namely:

- the original dbf (from the shapefiles) have a number of built-in attributes for the same features, for example series, time series, age group, sex group, urban/rural, along with yearly values.
- Multiple files can refer to the same indicator.
- Creating individual tables for each combination of the above-mentioned attributes would result in 795 views, which would show up as 795 layer in the GoeHub interface
- additionally, it is not possible to push performances in a meaningful way form the point of view of pg\_tileserv / PostgreSQL optimizations. It is not possible, for example, to dynamically join data to pre-semplified feature layers, because views can only join to a fixed table.

#### ​Examples

Indicator 10.5.1:

- 7 different files with the same indicator
- Layer name : Indicator 10.5.1 - Financial Soundness Indicators
- Available "time series":

| area | series_name |Desc |
| --- | --- | --- |
| assets to liabilities | FI\_FSI\_FSLS | Liquid\_assets\_to\_short\_term\_liabilities |
| foreign exchange | FI\_FSI\_FSSNO | Net\_open\_position\_in\_foreign\_exchange\_to\_capital |
| loans | FI\_FSI\_FSKNL | Non-performing\_loans\_net\_of\_provisions\_to\_capital |
| loans | FI\_FSI\_FSANL | Non-performing\_loans\_to\_total\_gross\_loans |
| capital to assets | FI\_FSI\_FSKA | Regulatory\_capital\_to\_assets |
| risk-weighted assets | FI\_FSI\_FSKRTC | Regulatory\_Tier\_1\_capital\_to\_risk-weighted\_assets |
| return on assets | FI\_FSI\_FSERA | Return\_on\_assets |

Indicator 4.6.1:

- 1 file

- 6 available "time series":

| area | series_name | age_group | sex_group |
| --- | --- | --- | --- |
| Literacy | SE\_ADT\_FUNS\_\_Y16T65\_\_\_T\_\_SKILL\_LTRCY | 16-65 years old | Total |
| Numeracy | SE\_ADT\_FUNS\_\_Y16T65\_\_\_T\_\_SKILL\_NMRCY | 16-65 years old | Total |
| Literacy | SE\_ADT\_FUNS\_\_Y16T65\_\_F\_\_SKILL\_LTRCY | 16-65 years old | Female |
| Numeracy | SE\_ADT\_FUNS\_\_Y16T65\_\_F\_\_SKILL\_NMRCY | 16-65 years old | Female |
| Literacy | SE\_ADT\_FUNS\_\_Y16T65\_\_M\_\_SKILL\_LTRCY | 16-65 years old | Male |
| Numeracy | SE\_ADT\_FUNS\_\_Y16T65\_\_M\_\_SKILL\_NMRCY | 16-65 years old | Male |

## ​Proposed implementation

The proposal would be therefore to create a set of subsettable functions to dynamically join geometry-less data with `admin0` features. Advantages:

- The number of layers added to GeoHub would be only ~240.
- Each layer would initially be presented with a default subset.
- Subsets are mutually exclusive (e.g.: either male or female, not both)
- Subsets are processed server-side, and are distinct from the already-implmented filters, which happen client-side.
- Layers could be subset by gender/age/urban area, etc.
- The subset options would extend back to the **existing function layers** and to the **tool piping**

![image](https://user-images.githubusercontent.com/18248460/225578335-3408a26a-d950-42fd-93ba-2f9434fe1942.png)

The concepts would be:

Dbf(s) →python code→ `sdgxx.admin0`

One PostgreSQL function per sdg code / indicator / series

Register & expose each PostgreSQL function as a vector layer (`{z}/{x}/{y}.pbf`)

Subsets for [time series] / sex group / sex age / urban / others (these will become WHEREs in PostgreSQL)

Pull-down menu or time slider to select the year (field format: `value_{yyyy}`)

The (partially) new feature would therefore be the presence of subsets.

Subset options will be exposed by the postgreSQL function via pg\_tileserv's endpoints as JSON (similarly to what we have today for the Subnational Dynamic HDI function):

[https://pgtileserv.undpgeohub.org/admin.dynamic\_subnational\_hdi/](https://pgtileserv.undpgeohub.org/admin.dynamic_subnational_hdi/)`{z}/{x}/{y}.pbf?params=`
`{"le_incr":{"value":11},"eys_incr":{"value":22},"mys_incr":{"value":33},"gni_incr":{"value":44}}`

will become:

[https://pgtileserv.undpgeohub.org/admin.dynamic\_subnational\_hdi/](https://pgtileserv.undpgeohub.org/admin.dynamic_subnational_hdi/)`{z}/{x}/{y}.pbf?params=`
`{"``values``":`
`{"le_incr":{"value":11},"eys_incr":{"value":22},"mys_incr":{"value":33},"gni_incr":{"value":44}},`
`"``subsets``":`
`{"sex_group":{"value":"Female","options":["Female","Male","Total"]},`
`"age_group":{"value":"18-24 years","options":["0-12 years","12-18 years","18-24 years","older than 24 years","total"]},`
`"location":{"value":"urban","options":["urban","rural","total"]}}}`

Please note that the current "params" dictionary argument would become `params→values`, and the subsets would be udner `params→subsets.`

Optional/future subset could be:

- by country/continent (to make queries faster, not filtered by the client) or
- by user-selected rectangle area.

**Comment by rafdouglas** [comment]
If a function layer exposes an empty `params→values`, or has no `params→values`, then the `simulation` tab should not be displayed.

Linked with https://github.com/UNDP-Data/geohub/issues/1558

**Reply by iferencik** [reply]
Correct
